File: java/app-encryption/src/main/java/com/godaddy/asherah/appencryption/SuffixedPartition.java
package com.godaddy.asherah.appencryption;

public class SuffixedPartition extends Partition {

  private final String regionSuffix;

  /**
   * Create a new {@code SuffixedPartition} instance using the provided parameters.
   *
   * @param partitionId A unique identifier for a {@link Session}.
   * @param serviceId A unique identifier for a service, used to create a {@link SessionFactory} object.
   * @param productId A unique identifier for a product, used to create a {@link SessionFactory} object.
   * @param regionSuffix A key suffix that prevents multi-region writes from clobbering each other and
   * ensures that no keys are lost.
   */
  public SuffixedPartition(final String partitionId, final String serviceId, final String productId,
                           final String regionSuffix) {
    super(partitionId, serviceId, productId);
    this.regionSuffix = regionSuffix;
  }

  @Override
  public String getSystemKeyId() {
    return super.getSystemKeyId() + "_" + this.regionSuffix;
  }

  @Override
  public String getIntermediateKeyId() {
    return super.getIntermediateKeyId() + "_" + this.regionSuffix;
  }

  @Override
  public String toString() {
    return getClass().getSimpleName() + "[partitionId=" + getPartitionId() +
      ", serviceId=" + getServiceId() + ", productId=" + getProductId() + ", regionSuffix=" + this.regionSuffix + "]";
  }

  @Override
  public boolean isValidIntermediateKeyId(final String id) {
    return id.equals(getIntermediateKeyId()) || id.startsWith(super.getIntermediateKeyId());
  }
}

File: csharp/AppEncryption/AppEncryption/SuffixedPartition.cs
namespace GoDaddy.Asherah.AppEncryption
{
    /// <summary>
    /// An implementation of <see cref="Partition"/> that is used to support Global Tables in
    /// <see cref="GoDaddy.Asherah.AppEncryption.Persistence.DynamoDbMetastoreImpl"/>.
    /// </summary>
    public class SuffixedPartition : Partition
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="SuffixedPartition"/> class.
        /// </summary>
        /// <param name="partitionId">A unique identifier for a <see cref="Session{TP,TD}"/>.</param>
        /// <param name="serviceId">A unique identifier for a service, used to create a <see cref="SessionFactory"/>
        /// object.</param>
        /// <param name="productId">A unique identifier for a product, used to create a <see cref="SessionFactory"/>
        /// object.</param>
        /// <param name="regionSuffix">The suffix to be added to a lookup key when using DynamoDB Global Tables.</param>
        public SuffixedPartition(string partitionId, string serviceId, string productId, string regionSuffix)
            : base(partitionId, serviceId, productId, regionSuffix)
        {
        }

        /// <inheritdoc/>
        public override string ToString()
        {
            return GetType().Name + "[partitionId=" + PartitionId +
                   ", serviceId=" + ServiceId + ", productId=" + ProductId + ", regionSuffix=" + Suffix + "]";
        }
    }
}
